embassy-mcxa/src/wwdt.rs

```diff
@@ -0,0 +1,220 @@
+pub struct Watchdog<'d> {
+    _peri: Peri<'d, WWDT0>,
+    load_value: u32,
+    frequency: u32,
+    // The register block of the WWDT instance
+    info: &'static pac::wwdt0::RegisterBlock,
+}
+
+impl<'d> Watchdog<'d> {
+    /// Create a new WWDT instance.
+    ///
+    /// # Arguments
+    ///
+    /// * `_peri` - The WWDT0 peripheral instance
+    /// * `_irq` - Interrupt binding for WWDT0
+    pub fn new(
+        _peri: Peri<'d, WWDT0>,
+        _irq: impl crate::interrupt::typelevel::Binding<typelevel::WWDT0, InterruptHandler> + 'd,
+    ) -> Self {
+        let _frequency = crate::clocks::with_clocks(|clocks| {
+            // Ensure clk_1m is active at the required power level
+            clocks.ensure_clk_1m_active(&crate::clocks::PoweredClock::NormalEnabledDeepSleepDisabled)
+        })
+        .expect("Clocks not initialized")
+        .expect("clk_1m not enabled or not at required power level");
+
+        // Enable WATCHDOG clock by writing to mrcc register
+        // Can't use enable_and_reset API here because WWDT doesn't have a reset signal.
+        let mrcc = unsafe { pac::Mrcc0::steal() };
+        mrcc.mrcc_glb_cc0().modify(|_, w| w.wwdt0().enabled());
+
+        crate::pac::Interrupt::WWDT0.unpend();
+        unsafe {
+            crate::pac::Interrupt::WWDT0.enable();
+        }
+        Self {
+            _peri,
+            load_value: 0,
+            frequency: _frequency / 4,
+            info: unsafe { &*pac::Wwdt0::ptr() },
+        }
+    }
+
+    // Start the watchdog timer with the specified timeout period.
+    ///
+    /// # Arguments
+    ///
+    /// * `period` - The timeout period after which the watchdog will trigger
+    /// * `warning` - Optional warning interrupt period before timeout. If `Some`, an interrupt
+    ///               will be generated at this time before the actual timeout. If `None`, the
+    ///               watchdog will directly reset the system on timeout.
+    ///
+    pub fn start(&mut self, period: Duration, warning: Option<Duration>) {
+        let timeout_cycles = (self.frequency as u64 * period.as_micros()) / 1_000_000;
+
+        // Ensure the value fits in u32 and is within valid range
+        // Writing a value below FFh causes 00_00FFh to load into the register. Therefore, the minimum timeout interval is TWDCLK X 256 X 4.
+        assert!(timeout_cycles <= 0xFFFFFF, "Timeout too large for watchdog counter");
+        assert!(timeout_cycles > 0xFF, "Timeout too small for watchdog counter");
+
+        self.load_value = timeout_cycles as u32;
+        self.set_timeout_value(self.load_value);
+
+        //Windows value is set to max at reset for no effect.
+
+        if let Some(warning_value) = warning {
+            let warning_cycles = (self.frequency as u64 * warning_value.as_micros()) / 1_000_000;
+            assert!(
+                warning_cycles <= 0x3FF,
+                "Warning value too large. At 1MHz, warning value must be <= 4.092ms."
+            );
```

Comment:
in general terms, these asserts should be returned to the caller as an `Err` variant of the `Result` type. `assert` will cause a `panic!()`.